Add lookups

LukePearson1 opened the pull request:
Resolves #41

LukePearson1 committed:
Port lookup tables

LukePearson1 committed:
Port multisets

LukePearson1 committed:
Port multisets

stechu commented on src/lookup/lookup.rs:
```diff
@@ -0,0 +1,74 @@
+impl StandardComposer {
+    /// Adds a plookup gate to the circuit with its corresponding 
+    /// constraints.
+    ///
```
would you mind writing the gate equation explicitly here?

LukePearson1 replied:
For now this has been removed @stechu. Only because lookups will be added in different PR's.

LukePearson1 committed:
Fix compilation

LukePearson1 committed:
Implement lookups

LukePearson1 committed:
Restore files

LukePearson1 committed:
Merge branch 'master' into add_lookups

LukePearson1 committed:
Apply cargo fmt

LukePearson1 committed:
Import polynomial

LukePearson1 committed:
Refactor multiset

LukePearson1 committed:
Merge branch 'master' into add_lookups

LukePearson1 committed:
Add lookups

lopeetall committed:
Added lookup table tests for both fields and fmt

lopeetall committed:
Added multiset tests for both fields and fmt

lopeetall committed:
Added witness table tests for both fields

lopeetall committed:
Added table preprocessing test for both fields

mathcrypto commented on plonk-core/src/lookup/lookup_table.rs:
```diff
@@ -0,0 +1,298 @@
+        self.0.push([F::from(a), F::from(b), F::from(c), F::zero()]);
+    }
+
+    /// Insert a new row for an addition operation.
```
insert a row for a special operation not addition

LukePearson1 committed:
Merge branch 'master' into add_lookups

LukePearson1 committed:
Merge branch 'master' into add_lookups

LukePearson1 committed:
Merge branch 'master' into add_lookups

LukePearson1 committed:
Include Rand crate

LukePearson1 committed:
Refactor prover to include lookups (#84)

* Refactor prover to include lookups

This is the first commit to implement the
plonkup proving system.

* Reference Scalar

* Lookup error fixing (#87)

* Include PairingEngine in impls calling ProverKey

* Added lookup fields and derefed where needed

* Fixed some poly math

Co-authored-by: Joshua Fitzgerald <joshuabfitzgerald@gmail.com>

* append challenges to transcript

* assert challenges  different

* compute permutation challenges

* add missing challenges commitments

* Fix referenced variables

* Partial refactor

This commit contains various fixes
to the proving algorithm to implement
trait based features

* Remove some errors

* Refactor proving algorithm

* Finalise prover algorithm

* removed unneeded q_lookup padding; fixed some comments

Co-authored-by: Joshua Fitzgerald <joshuabfitzgerald@gmail.com>
Co-authored-by: mathcrypto <bouguera_amira@yahoo.fr>
